File: ios/background_geo_tracker/Sources/background_geo_tracker/PointQueue.swift
import Foundation

/// The durable point queue. Enforces both ceilings on every write so the
/// database stays bounded even if uploading has been failing for days.
final class PointQueue {
    private let store: PointStore

    init(store: PointStore) {
        self.store = store
    }

    /// The one queue everything shares.
    ///
    /// The tracker writes while the uploader deletes, and two independent
    /// connections to the same file would leave that overlap to SQLite's file
    /// locking — losing writes to SQLITE_BUSY. One connection, one lock.
    static let shared: PointQueue? = PointStore(path: defaultPath())
        .map(PointQueue.init)

    /// The on-device location, alongside the app's other support files.
    static func defaultPath() -> String {
        let directory = FileManager.default.urls(
            for: .applicationSupportDirectory, in: .userDomainMask
        )[0]
        try? FileManager.default.createDirectory(
            at: directory, withIntermediateDirectories: true
        )
        return directory.appendingPathComponent("attractor_geo.sqlite").path
    }

    @discardableResult
    func enqueue(
        _ point: GeoPointRow, maxPoints: Int, maxAgeDays: Int
    ) -> Bool {
        guard store.insert(point) else { return false }
        // The cutoff is measured from the incoming point rather than the wall
        // clock, so a device whose clock jumped cannot wipe a valid queue.
        let window = Int64(maxAgeDays) * 86_400_000
        store.deleteOlderThan(point.recordedAtMillis - window)
        if store.count() > maxPoints {
            store.trim(to: maxPoints)
        }
        return true
    }

    func oldest(limit: Int) -> [GeoPointRow] { store.oldest(limit: limit) }

    func drop(ids: [String]) { store.delete(ids: ids) }

    func count() -> Int { store.count() }

    /// Throws the whole queue away. For signing out: these points belong to
    /// whoever recorded them, and must not be uploaded by the next account on
    /// this device.
    func clear() { store.deleteAll() }
}

File: android/src/main/kotlin/dev/background_geo_tracker/plugin/db/PointQueue.kt
package dev.background_geo_tracker.plugin.db

/**
 * The durable point queue. Enforces both ceilings on every write so the
 * database stays bounded even if uploading has been failing for days.
 */
class PointQueue(private val dao: PointDao) {

    fun enqueue(point: PointRow, maxPoints: Int, maxAgeDays: Int) {
        dao.insert(point)
        // The cutoff is measured from the incoming point rather than the wall
        // clock, so a device whose clock jumped cannot wipe a valid queue.
        dao.deleteOlderThan(point.recordedAtMillis - maxAgeDays * DAY_MILLIS)
        if (dao.count() > maxPoints) {
            dao.trimTo(maxPoints)
        }
    }

    fun oldest(limit: Int, nowMillis: Long): List<PointRow> =
        dao.oldest(limit, nowMillis)

    /** Stands a refused batch down so the queue behind it can move. */
    fun defer(ids: List<String>, untilMillis: Long) =
        dao.defer(ids, untilMillis)

    fun oldestForSession(sessionId: String, limit: Int): List<PointRow> =
        dao.oldestForSession(sessionId, limit)

    fun drop(ids: List<String>) = dao.deleteByIds(ids)

    fun count(): Int = dao.count()

    fun countForSession(sessionId: String): Int = dao.countForSession(sessionId)

    /**
     * Throws the whole queue away. For signing out: these points belong to
     * whoever recorded them, and must not be uploaded by the next account on
     * this device.
     */
    fun clear() = dao.deleteAll()

    private companion object {
        const val DAY_MILLIS = 86_400_000L
    }
}
